Static condensation followed by field split

lindsayad opened the pull request:
This implements capability in MOOSE that before I'd only been able to exercise through firedrake: do field splits on a statically condensed system. The idea is to eventually apply this to interior penalty HDG Navier Stokes in which we statically condense out the internal velocity and pressure degrees of freedom, and then do a field split between the velocity and pressure facet dofs

lindsayad committed:
Allow doing field split on statically condensed systems

Closes #30798

- Move field split/preconditioning tasks after default convergence.
  This somehow makes the SAM tests pass. I have no idea
- Fix mass matrix integration for HDG
- Move user object task before preconditioning

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2999579/), step Docs: sync website on 5e0275f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30357/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3003404/), step Generate coverage on 7a6def4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>5e2432</th>
      <th colspan=4>#30357 7a6def</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.48%</td>
      <td>+0.00%</td>
      <td>86.81%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>114883</td>
      <td>114978</td>
      <td>+95</td>
      <td>204</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19525</td>
      <td>19534</td>
      <td>+9</td>
      <td>31</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30357/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30357/coverage/framework)


## Modules coverage

### Fsi

<table>
  <thead>
    <tr>
      <th></th>
      <th>5e2432</th>
      <th colspan=4>#30357 7a6def</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.56%</td>
      <td>88.53%</td>
      <td>-0.03%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>333</td>
      <td>332</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>43</td>
      <td>43</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30357/coverage_diff/fsi)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30357/coverage/fsi)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>5e2432</th>
      <th colspan=4>#30357 7a6def</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.24%</td>
      <td>85.29%</td>
      <td>+0.05%</td>
      <td>95.37%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>18785</td>
      <td>18857</td>
      <td>+72</td>
      <td>103</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3252</td>
      <td>3252</td>
      <td>-</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30357/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30357/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30357/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 86.81% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3008822/) on e9454e8 : invalidated by @lindsayad

moosebuild commented:
Job [Disable HDF5](https://civet.inl.gov/job/3029378/) on e37be76 : invalidated by @lindsayad

moosebuild commented:
Job [No optional](https://civet.inl.gov/job/3029380/) on e37be76 : invalidated by @lindsayad

moosebuild commented:
Job [Non unity build](https://civet.inl.gov/job/3029383/) on e37be76 : invalidated by @lindsayad

GiudGiud commented on modules/navier_stokes/doc/content/source/hdgkernels/MassFluxPenaltyIPHDG.md:
```diff
@@ -0,0 +1,30 @@
+# MassFluxPenaltyIPHDG
+
+Unlike the perturbation in [MassFluxPenalty.md], this class does not introduce coupling between interior degress of freedom on adjacent elements. Instead it imposes the face term
```
```suggestion
Unlike the perturbation in [MassFluxPenalty.md], this class does not introduce coupling between interior degress of freedom on adjacent elements. Instead it imposes the element face term
```

GiudGiud commented on modules/navier_stokes/doc/content/source/hdgkernels/MassFluxPenaltyIPHDG.md:
```diff
@@ -0,0 +1,30 @@
+symmetric, singular $n \times n$ matrix, $B$ is an $m \times n$
+matrix ($m < n$), and $\gamma>0$ is a scalar
+constant. Requirements for the singular perturbation are that $\mathcal{N}(J)\subseteq \mathcal{N}(B)$
+and $\mathcal{R}(B^T) = \mathcal{R}(J)$. [eq:perturbation] meets these requirements.
```
maybe to be more clear, this term forms the B equations?

GiudGiud commented on modules/navier_stokes/doc/content/source/hdgkernels/MassFluxPenaltyIPHDG.md:
```diff
@@ -0,0 +1,30 @@
+# MassFluxPenaltyIPHDG
```
should probably mention that this goes in the momentum equations not the mass one

GiudGiud replied:
even though it s called Mass

GiudGiud commented on modules/navier_stokes/include/hdgkernels/MassFluxPenaltyIPHDG.h:
```diff
@@ -0,0 +1,53 @@
+
+#include "HDGKernel.h"
+
+class MassFluxPenaltyIPHDG : public HDGKernel
```
docstring

GiudGiud commented on modules/navier_stokes/include/hdgkernels/MassFluxPenaltyIPHDG.h:
```diff
@@ -0,0 +1,53 @@
+  std::vector<ADReal> _ad_residuals;
+
+private:
+  template <typename T>
```
docstring

GiudGiud commented on framework/include/preconditioners/FieldSplitPreconditioner.h:
```diff
@@ -9,30 +9,94 @@
+  virtual void setupDM() = 0;
+
+  /**
+   * @returns The KSP object associated with the field split preconditioner
```
or is this just for the field split. in which case not all field split types use it right?

lindsayad replied:
there is one outer KSP for any field split preconditioner. Then there are inner KSPs as well for the various splits

GiudGiud reviewed (commented):
sorry for the delay!

lindsayad commented on framework/include/preconditioners/FieldSplitPreconditioner.h:
```diff
@@ -9,30 +9,94 @@
+  virtual const libMesh::System & system() const = 0;
+
+  /**
+   * @returns The prefix to pass to PETSc for the DM
```
that would probably be confusing to some people as this class holds a `_nl` and the prefix passed to PETSc for static condensation is not the `_nl` system's prefix

lindsayad commented on framework/include/preconditioners/StaticCondensationFieldSplitPreconditioner.h:
```diff
@@ -0,0 +1,36 @@
+  : public FieldSplitPreconditionerTempl<MooseStaticCondensationPreconditioner>
+{
+public:
+  /**
+   *  Constructor. Initializes SplitBasedPreconditioner data structures
+   */
```
wrekt

lindsayad committed:
Cleanup prefixing to be less confusing

lindsayad commented on framework/include/systems/SystemBase.h:
```diff
@@ -962,10 +962,21 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+  std::string petscPrefix() const;
+
+  /**
+   * size the matrix data for each variable
```
yea, will expand

lindsayad committed:
Matrix tag cleanups

- Better doxygen
- active -> activate
- Remove unused singleton activateMatrixTag/deactivateMatrixTag

lindsayad commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/lid-driven/gold/lid-driven-scfsp_out.e:
With the cli args this only has one timestep. Are you saying don't execute on initial? I like to have the 2D image since it shows the asymmetry for advection dominated

lindsayad committed:
Add back a paramError check for WallDistanceMixingLengthAux

lindsayad commented on framework/src/preconditioners/FieldSplitPreconditioner.C:
```diff
@@ -40,27 +45,34 @@ FieldSplitPreconditioner::validParams()
+    _nl(this->_fe_problem.getNonlinearSystemBase(this->_nl_sys_num)),
+    _decomposition_split(this->template getParam<std::string>("topsplit"))
 {
+  if (libMesh::default_solver_package() != libMesh::PETSC_SOLVERS)
+    mooseError("The field split preconditioner can only be used with PETSc");
+
```
this is just an additional sanity check. I can remove it if you prefer

lindsayad commented on framework/src/preconditioners/FieldSplitPreconditioner.C:
```diff
@@ -40,27 +45,34 @@ FieldSplitPreconditioner::validParams()
   // it is recommended to have a full Jacobian for using
   // the fieldSplit preconditioner
-  bool full = getParam<bool>("full");
+  bool full = this->template getParam<bool>("full");
```
Users can specify off-diagonal blocks when `!full`, so it doesn't mean just diagonals. For additive splitting, you only need/use block diagonals. For multiplicative you would need the lower triangular blocks. For Schur you do need full

lindsayad commented on modules/navier_stokes/src/problems/NavierStokesProblem.C:
```diff
@@ -367,9 +389,7 @@ NavierStokesProblem::initPetscOutputAndSomeSolverSettings()
-  KSP ksp;
-  auto snes = currentNonlinearSystem().getSNES();
-  LibmeshPetscCall(SNESGetKSP(snes, &ksp));
+  auto ksp = currentNonlinearSystem().getFieldSplitPreconditioner().getKSP();
```
Yea this class only works with field split preconditioners. It says in the doc string that the class is for Schur complement preconditioning. There are various `mooseError` throughout the class related to checking for FSP

lindsayad committed:
Improve documentation

lindsayad committed:
Reduce gold exodus size

Don't include initial conditions

moosebuild commented:
Job [Non unity build](https://civet.inl.gov/job/3094452/) on cced0b4 : invalidated by @lindsayad

GiudGiud reviewed (commented):
getting trouble with block restriction. Maybe a user error though?
sending you an input

lindsayad commented on modules/navier_stokes/src/executioners/SIMPLESolveBase.C:
```diff
@@ -341,12 +341,9 @@ SIMPLESolveBase::SIMPLESolveBase(Executioner & ex)
-                                                   "-",
-                                                   *this,
-                                                   _momentum_petsc_options);
+      momentum_petsc_options, "", *this, _momentum_petsc_options);
```
PETSc APIs for setting prefixes do not take dashes. The time that you do need a dash is when calling `PetscOptionsSetValue` if you are programmatically setting petsc options, as opposed to passing the petsc options directly from the command line

lindsayad committed:
Add dof_map doxygen

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Add block restriction test

lindsayad commented:
will wait on libmesh/libmesh#4224

lindsayad committed:
Fix naming of AddTaggedMatricesAction

lindsayad committed:
Remove facet variables from output
